src/types.jl

```diff
@@ -75,12 +96,16 @@ end
 (::Type{D})(::Type{R}; kws...) where {R,D<:AbstractDimensions} = constructor_of(D){R}((tryrationalize(R, get(kws, k, zero(R))) for k in static_fieldnames(D))...)
 (::Type{D})(; kws...) where {R,D<:AbstractDimensions{R}} = constructor_of(D)(R; kws...)
 (::Type{D})(; kws...) where {D<:AbstractDimensions} = D(DEFAULT_DIM_BASE_TYPE; kws...)
-(::Type{D})(d::AbstractDimensions) where {R,D<:AbstractDimensions{R}} = D((getproperty(d, k) for k in static_fieldnames(D))...)
+function (::Type{D})(d::D2) where {R,D<:AbstractDimensions{R},D2<:AbstractDimensions}
+    issetequal(static_fieldnames(D), static_fieldnames(D2)) ||
```

Comment:
This check seems to have caused the regression on 1.9; I'm not sure exactly why, but presumably the line isn't running entirely at compile time as we would expect on 1.9?

Reply:
Good catch! Maybe `==` is the right operator here…

Reply:
Okay I fixed it with a generated function for the fieldnames equality check. I guess 1.9 just didn't want to inline it.